PlayerAgent.cs

```diff
@@ -41,77 +63,116 @@ public class PlayerAgent : Agent
+        orginalTargetPosition = target.transform.localPosition; 
     }
 
     public override void OnEpisodeBegin()
     {
         transform.LookAt(target.transform);
+
         target.transform.localPosition = orginalTargetPosition;
+
         transform.localPosition = orginalPosition;
-        transform.localPosition = new Vector3(orginalPosition.x, orginalPosition.y, Random.Range(22,24));
-        nextStepTimeout = StepCount + stepTimeout;
 
+        transform.localPosition = new Vector3(orginalPosition.x, orginalPosition.y, Random.Range(22f, 24f));
+
+        // nextStepTimeout = StepCount + stepTimeout; // we don't have to keep checkpoints
+
+
+        // Measure the distance from target at start
+        distanceAtStart = Vector3.Distance(transform.localPosition, orginalTargetPosition);
+
+        // reset player velocity
+        playerRigidbody.velocity = Vector3.zero;
+
+        //lastPlayerPosition = transform.localPosition;
+       
     }
 
     public override void CollectObservations(VectorSensor sensor)
     {
         // 3 obserwacje
         sensor.AddObservation(transform.localPosition);
+
         // 3 obserwacje
         sensor.AddObservation(target.transform.localPosition);
+
         // 1 obserwacja
         sensor.AddObservation(playerRigidbody.velocity.x);
+
         // 1 obserwacja
         sensor.AddObservation(playerRigidbody.velocity.z);
 
         // 1 obserwacja
         sensor.AddObservation(Vector3.Distance(transform.localPosition, target.transform.localPosition));
-
-
     }
 
     public override void OnActionReceived(ActionBuffers actions)
     {
+
+        //var lastDistanceFromTarget = Vector3.Distance(lastPlayerPosition, target.transform.localPosition);
+        Debug.Log("distanceAtStart " + distanceAtStart);
+
+        // Actions
         var vectorForce = new Vector3();
+
         vectorForce.x = actions.ContinuousActions[0];
+
         vectorForce.z = actions.ContinuousActions[1];
-        
 
         playerRigidbody.AddForce(vectorForce * speed);
+        //Debug.Log("currentDistanceFromTarget " + lastDistanceFromTarget);
 
-        var distanceFromTarget = Vector3.Distance(transform.localPosition, target.transform.localPosition);
+        //lastPlayerPosition = transform.localPosition;
+        var currentDistanceFromTarget = Vector3.Distance(transform.localPosition, target.transform.localPosition);
+        Debug.Log("lastDistance " + currentDistanceFromTarget);
 
+        // Reward function
+        var reward = (float)score(currentDistanceFromTarget);
+        // Set reward once
+        SetReward(reward);
 
-        // Nagradzanie agenta
+        //Debug.Log("reward "+ reward);
 
-        AddReward(-1f / MaxStep);
+        // if (currentDistanceFromTarget < lastDistanceFromTarget) //good condition?
+        if (currentDistanceFromTarget > distanceAtStart)
+        {
+            AddReward(-0.1f);
+            StartCoroutine(SwapGroundMaterial(goAwayMaterial, 0.5f));
+            EndEpisode();
```

Comment:
Gdy currentDistance jest większy niż początkowa odległość między playerem a targetem to podłoga świeci się na złoto i kończymy episod jak nie zakończymy to kule wylatują w powietrze a wszystkie zmienne distance się zwiększają